File: android/src/main/java/ai/synheart/behavior/StatsCollector.kt
package ai.synheart.behavior

import java.util.concurrent.ConcurrentHashMap
import java.util.LinkedList

/**
 * Collects and maintains rolling statistics for behavioral signals.
 */
class StatsCollector {

    private val recentEvents = LinkedList<BehaviorEvent>()
    private val maxEvents = 100

    // Rolling metrics
    private var latestTypingCadence: Double? = null
    private var latestInterKeyLatency: Double? = null
    private var latestBurstLength: Int? = null
    private var latestScrollVelocity: Double? = null
    private var latestScrollAcceleration: Double? = null
    private var latestScrollJitter: Double? = null
    private var latestTapRate: Double? = null
    private var latestForegroundDuration: Double? = null
    private var latestIdleGapSeconds: Double? = null
    private var latestStabilityIndex: Double? = null
    private var latestFragmentationIndex: Double? = null

    private val appSwitchTimestamps = LinkedList<Long>()

    @Synchronized
    fun recordEvent(event: BehaviorEvent) {
        recentEvents.add(event)
        while (recentEvents.size > maxEvents) {
            recentEvents.removeFirst()
        }

        // Update metrics based on event type
        when (event.type) {
            "typingCadence" -> {
                latestTypingCadence = event.payload["cadence"] as? Double
                latestInterKeyLatency = event.payload["inter_key_latency"] as? Double
            }
            "typingBurst" -> {
                latestBurstLength = event.payload["burst_length"] as? Int
                latestInterKeyLatency = event.payload["inter_key_latency"] as? Double
            }
            "scrollVelocity" -> {
                latestScrollVelocity = event.payload["velocity"] as? Double
            }
            "scrollAcceleration" -> {
                latestScrollAcceleration = event.payload["acceleration"] as? Double
            }
            "scrollJitter" -> {
                latestScrollJitter = event.payload["jitter"] as? Double
            }
            "tapRate" -> {
                latestTapRate = event.payload["tap_rate"] as? Double
            }
            "foregroundDuration" -> {
                latestForegroundDuration = (event.payload["duration_seconds"] as? Number)?.toDouble()
            }
            "idleGap" -> {
                latestIdleGapSeconds = (event.payload["idle_seconds"] as? Number)?.toDouble()
            }
            "sessionStability" -> {
                latestStabilityIndex = event.payload["stability_index"] as? Double
                latestFragmentationIndex = event.payload["fragmentation_index"] as? Double
            }
            "appSwitch" -> {
                appSwitchTimestamps.add(event.timestamp)
                // Keep only last minute
                val cutoff = System.currentTimeMillis() - 60000
                while (appSwitchTimestamps.isNotEmpty() && appSwitchTimestamps.first() < cutoff) {
                    appSwitchTimestamps.removeFirst()
                }
            }
        }
    }

    @Synchronized
    fun getCurrentStats(): BehaviorStats {
        return BehaviorStats(
            typingCadence = latestTypingCadence,
            interKeyLatency = latestInterKeyLatency,
            burstLength = latestBurstLength,
            scrollVelocity = latestScrollVelocity,
            scrollAcceleration = latestScrollAcceleration,
            scrollJitter = latestScrollJitter,
            tapRate = latestTapRate,
            appSwitchesPerMinute = appSwitchTimestamps.size,
            foregroundDuration = latestForegroundDuration,
            idleGapSeconds = latestIdleGapSeconds,
            stabilityIndex = latestStabilityIndex,
            fragmentationIndex = latestFragmentationIndex,
            timestamp = System.currentTimeMillis()
        )
    }

    @Synchronized
    fun clear() {
        recentEvents.clear()
        appSwitchTimestamps.clear()
        latestTypingCadence = null
        latestInterKeyLatency = null
        latestBurstLength = null
        latestScrollVelocity = null
        latestScrollAcceleration = null
        latestScrollJitter = null
        latestTapRate = null
        latestForegroundDuration = null
        latestIdleGapSeconds = null
        latestStabilityIndex = null
        latestFragmentationIndex = null
    }
}

File: ios/Classes/StatsCollector.swift
import Foundation

/// Collects and maintains rolling statistics for behavioral signals.
class StatsCollector {

    private var recentEvents: [BehaviorEvent] = []
    private let maxEvents = 100
    private let queue = DispatchQueue(label: "ai.synheart.behavior.stats", attributes: .concurrent)

    // Rolling metrics
    private var latestTypingCadence: Double?
    private var latestInterKeyLatency: Double?
    private var latestBurstLength: Int?
    private var latestScrollVelocity: Double?
    private var latestScrollAcceleration: Double?
    private var latestScrollJitter: Double?
    private var latestTapRate: Double?
    private var latestForegroundDuration: Double?
    private var latestIdleGapSeconds: Double?
    private var latestStabilityIndex: Double?
    private var latestFragmentationIndex: Double?

    private var appSwitchTimestamps: [Double] = []

    func recordEvent(_ event: BehaviorEvent) {
        queue.async(flags: .barrier) {
            self.recentEvents.append(event)
            if self.recentEvents.count > self.maxEvents {
                self.recentEvents.removeFirst()
            }

            // Update metrics based on event type
            switch event.type {
            case "typingCadence":
                self.latestTypingCadence = event.payload["cadence"] as? Double
                self.latestInterKeyLatency = event.payload["inter_key_latency"] as? Double

            case "typingBurst":
                self.latestBurstLength = event.payload["burst_length"] as? Int
                self.latestInterKeyLatency = event.payload["inter_key_latency"] as? Double

            case "scrollVelocity":
                self.latestScrollVelocity = event.payload["velocity"] as? Double

            case "scrollAcceleration":
                self.latestScrollAcceleration = event.payload["acceleration"] as? Double

            case "scrollJitter":
                self.latestScrollJitter = event.payload["jitter"] as? Double

            case "tapRate":
                self.latestTapRate = event.payload["tap_rate"] as? Double

            case "foregroundDuration":
                if let durationSeconds = event.payload["duration_seconds"] as? Double {
                    self.latestForegroundDuration = durationSeconds
                } else if let durationMs = event.payload["duration_ms"] as? Double {
                    self.latestForegroundDuration = durationMs / 1000.0
                }

            case "idleGap":
                if let idleSeconds = event.payload["idle_seconds"] as? Double {
                    self.latestIdleGapSeconds = idleSeconds
                }

            case "sessionStability":
                self.latestStabilityIndex = event.payload["stability_index"] as? Double
                self.latestFragmentationIndex = event.payload["fragmentation_index"] as? Double

            case "appSwitch":
                let timestamp = Double(event.timestamp)
                self.appSwitchTimestamps.append(timestamp)

                // Keep only last minute
                let cutoff = Date().timeIntervalSince1970 * 1000 - 60000
                self.appSwitchTimestamps = self.appSwitchTimestamps.filter { $0 >= cutoff }

            default:
                break
            }
        }
    }

    func getCurrentStats() -> BehaviorStats {
        return queue.sync {
            return BehaviorStats(
                typingCadence: latestTypingCadence,
                interKeyLatency: latestInterKeyLatency,
                burstLength: latestBurstLength,
                scrollVelocity: latestScrollVelocity,
                scrollAcceleration: latestScrollAcceleration,
                scrollJitter: latestScrollJitter,
                tapRate: latestTapRate,
                appSwitchesPerMinute: appSwitchTimestamps.count,
                foregroundDuration: latestForegroundDuration,
                idleGapSeconds: latestIdleGapSeconds,
                stabilityIndex: latestStabilityIndex,
                fragmentationIndex: latestFragmentationIndex,
                timestamp: Int64(Date().timeIntervalSince1970 * 1000)
            )
        }
    }

    func clear() {
        queue.async(flags: .barrier) {
            self.recentEvents.removeAll()
            self.appSwitchTimestamps.removeAll()
            self.latestTypingCadence = nil
            self.latestInterKeyLatency = nil
            self.latestBurstLength = nil
            self.latestScrollVelocity = nil
            self.latestScrollAcceleration = nil
            self.latestScrollJitter = nil
            self.latestTapRate = nil
            self.latestForegroundDuration = nil
            self.latestIdleGapSeconds = nil
            self.latestStabilityIndex = nil
            self.latestFragmentationIndex = nil
        }
    }
}
